cts/CTS_Design_Document.md

```diff
@@ -0,0 +1,341 @@
+```python
+def set_anari_severity(severity)
+```
+
+## Features
+
+### Render a set of known test scenes
+#### Define test scene format
+
+Internally, use a dictionary for all the parameter settings. This could be achieved e.g. via JSON. To generate scenes, some parameters need to be specified, others should be consistent for all scenes for structural testing. Some parameters which should not change could entail: camera position, frame resolution, light settings (e.g. uniform white HDR) and used material.
+The test scene are generated from code without needing to load external files. The bounds properties and maximum Euclidean distance need to be generated as well to create a ground truth to test against. The test scenes should be generated with defined parameter permutations e.g. a triangle should be rendered using a triangle soup and using an index buffer. All ANARI geometries should be tested. For triangles and quads simple composite geometries are tested as well (Plane, Cube). The test scenes should include a list of ANARI features which are necessary to run the test. This can be used to check if required features are implemented on the ANARI device and automatically skip these tests if necessary. Additionally, the test scenes should be able to cover edge cases. The ANARI SDK already implements a scene framework in [libs/anari_test_scenes](../libs/anari_test_scenes/). This could be extended to allow the generation of scenes.
+#### Python API
+
+The Python API is used by the user to invoke the rendering. The function could look similar to this:
+```python
+def render_scenes(anari_library, anari_device = None, anari_renderer = "default", test_scenes = "all", output = ".")
+```
+This invokes the required ANARI calls for the specified device and library in the C++ backend via pybind11. If `anari_device` is set to `None`, the default device is used (first device of anariGetDeviceSubtypes). Now, all specified test scenes are rendered with the specified renderer. If no renderer is given, the default renderer is used. `test_scenes` can either be a category denoted by a string or be a number of scenes denoted by a list of strings. If no scene is given, all test scenes are used by default. `output` specifies the folder in which the renderings are saved. It defaults to the folder where the script is executed from. The python API receives the pixel data from the C++ backend and writes the renderings to disk.
```

Comment:
I'm a bit confused on the "string" set of scenes.  If the scenes are defined as python functions that generate them, why not just a list / set of scene items?  

If we need to filter that list of scenes by a category that can be done outside this function too.

Reply:
Maybe I'm misunderstanding here.  Are the tests staying in CPP or going to be made in python?  The actual code that generates the ANARI scene?

Reply:
The `test_scenes` parameter was intended to be used either as a category or a list of scenes. We can also add a helper function which resolves scene categories into a list of scenes to simplify the parameter. In general, I created the functions definitions more with args_parse/CLI in mind. So one can translate the function's parameters directly to arguments (e.g. `--test_scenes all` -> `test_scenes = "all"` without calling a helper function). But I can add a section specifically for args_parse, and we can simplify the actual python functions if that is desired.

Reply:
The scenes will probably be generated in C++, especially if the already existing scene system is used. From my understanding, the scenes should be generated during runtime with predefined parameters. So no new scene cpp files with vertex data etc. are added to the repo. But it is also fine, if we want to have a generator which creates these scene cpps (like [instanced_cubes](https://github.com/KhronosGroup/ANARI-SDK/blob/main/libs/anari_test_scenes/scenes/instanced_cubes.cpp)) once.

Reply:
I'd say pick one, scenes generated in python or C++.  If all the scenes are generated in python then just passing a list of functions that generate the scenes makes more sense to me.

Reply:
From my understanding, the scenes should be generated similar to [random_spheres](https://github.com/KhronosGroup/ANARI-SDK/blob/main/libs/anari_test_scenes/scenes/random_spheres.cpp). Therefore, we provide C++ classes which can generated different scenes based on the input parameters. The actual input parameters are statically defined in python (or maybe a json file), but this is done in another project. See the following quote from the RFP:
`The work outlined in this document only covers the establishment of such a test scene system: the ANARI
Working Group is responsible for defining the complete set of scenes that represent conformance of the ANARI
API. Work to generate a comprehensive list of tests may come in a future RFP.`

Another requirement from the RFP is the hierarchical structure of the test scenes:
`The ANARI CTS needs a command-line application that can render one or more test scenes that are arranged in
a hierarchy for logical grouping.`
This is the reason we added the parameter as a category. As I understood the project, it should mainly be used as a command line tool, and therefore passing actual functions is not possible and only predefined test scenes should be selected.

We can also discuss this more in-depth in the upcoming meeting on Thursday.